src/utils/common/common_test.c

```diff
@@ -299,53 +299,285 @@ DEF_TEST(strunescape) {
+      break;
+    case METRIC_TYPE_FPCOUNTER:
+      EXPECT_EQ_DOUBLE(cases[i].want.fpcounter, got.fpcounter);
+      EXPECT_EQ_UINT64(cases[i].want.fpcounter, state.last_value.fpcounter);
+      break;
+    case METRIC_TYPE_UNTYPED:
+      LOG(false, "invalid metric type");
+      break;
+    }
+
+    EXPECT_EQ_UINT64(cases[i].time, state.last_time);
+    EXPECT_EQ_DOUBLE(cases[i].want_residual, state.residual);
+  }
+
+  return 0;
+}
+
 DEF_TEST(value_to_rate) {
   struct {
+    char *name;
     time_t t0;
     time_t t1;
-    int ds_type;
+    metric_type_t type;
     value_t v0;
     value_t v1;
     gauge_t want;
+    int want_err;
   } cases[] = {
-      {0, 10, DS_TYPE_DERIVE, {.derive = 0}, {.derive = 1000}, NAN},
-      {10, 20, DS_TYPE_DERIVE, {.derive = 1000}, {.derive = 2000}, 100.0},
-      {20, 30, DS_TYPE_DERIVE, {.derive = 2000}, {.derive = 1800}, -20.0},
-      {0, 10, DS_TYPE_COUNTER, {.counter = 0}, {.counter = 1000}, NAN},
-      {10, 20, DS_TYPE_COUNTER, {.counter = 1000}, {.counter = 5000}, 400.0},
-      /* 32bit wrap-around. */
-      {20,
-       30,
-       DS_TYPE_COUNTER,
-       {.counter = 4294967238ULL},
-       {.counter = 42},
-       10.0},
-      /* 64bit wrap-around. */
-      {30,
-       40,
-       DS_TYPE_COUNTER,
-       {.counter = 18446744073709551558ULL},
-       {.counter = 42},
-       10.0},
+      {
+          .name = "derive_t init",
+          .t0 = 0,
+          .t1 = 10,
+          .type = DS_TYPE_DERIVE,
+          .v0 = {.derive = 0},
+          .v1 = {.derive = 1000},
+          .want_err = EAGAIN,
+      },
+      {
+          .name = "derive_t increase",
+          .t0 = 10,
+          .t1 = 20,
+          .type = DS_TYPE_DERIVE,
+          .v0 = {.derive = 1000},
+          .v1 = {.derive = 2000},
+          .want = 100.0,
+      },
+      {
+          .name = "derive_t decrease",
+          .t0 = 20,
+          .t1 = 30,
+          .type = DS_TYPE_DERIVE,
+          .v0 = {.derive = 2000},
+          .v1 = {.derive = 1800},
+          .want = -20.0,
+      },
+      {
+          .name = "counter_t init",
+          .t0 = 0,
+          .t1 = 10,
+          .type = METRIC_TYPE_COUNTER,
+          .v0 = {.counter = 0},
+          .v1 = {.counter = 1000},
+          .want_err = EAGAIN,
+      },
+      {
+          .name = "counter_t increase",
+          .t0 = 10,
+          .t1 = 20,
+          .type = METRIC_TYPE_COUNTER,
+          .v0 = {.counter = 1000},
+          .v1 = {.counter = 5000},
+          .want = 400.0,
+      },
+      {
+          .name = "counter_t 32bit wrap-around",
+          .t0 = 20,
+          .t1 = 30,
+          .type = METRIC_TYPE_COUNTER,
+          .v0 = {.counter = 4294967238ULL},
+          .v1 = {.counter = 42},
+          .want = 10.0,
+      },
+      {
+          .name = "counter_t 64bit wrap-around",
+          .t0 = 30,
+          .t1 = 40,
+          .type = METRIC_TYPE_COUNTER,
+          .v0 = {.counter = 18446744073709551558ULL},
```

Comment:
Wrap-around initial values would IMHO be clearer as `0xffffffff-57` and `0xffffffffffffffff-57`.